File: Week 2 - Arrays/Readability/readability.c
// Compute the coleman-liau index of a given input string.

#include <stdio.h>
#include <string.h>
#include <cs50.h>
#include <math.h>

int main(void)
{
    // Get input from user
    string input = get_string("Text: ");

    // return the number of letters in the input
    int letters = 0;
    for (int i = 0; i < strlen(input); i++)
    {
        if (input[i] >= 'a' && input[i] <= 'z')
        {
            letters++;
        }
        else if (input[i] >= 'A' && input[i] <= 'Z')
        {
            letters++;
        }
    }

    // return the number of words in the input
    int words = 1;
    for (int i = 0; i < strlen(input); i++)
    {
        if (input[i] == ' ')
        {
            words++;
        }
    }

    // return the number of sentences in the input
    int sentences = 0;
    for (int i = 0; i < strlen(input); i++)
    {
        if (input[i] == '.' || input[i] == '!' || input[i] == '?')
        {
            sentences++;
        }
    }

    // compute the Coleman-Liau index
    float L = (float) letters / (float) words * 100;
    float S = (float) sentences / (float) words * 100;
    float index = 0.0588 * L - 0.296 * S - 15.8;

    // if the index is less than 1, print "Before Grade 1"
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    // if the index is between 1 and 16, print "Grade %i"
    else if (index >= 1 && index <= 16)
    {
        printf("Grade %i\n", (int) round(index));
    }
    // if the index is 16 or higher, print "Grade 16+"
    else
    {
        printf("Grade 16+\n");
    }

    return 0;
}

File: Week 6 - Python/pset6/readability/readability.py
"""A program that computes the coleman-liau index of a given piece of text, received as an input."""

import re

TEXT = input("Text: ")


# return the number of letters in the text, excluding spaces
def count_letters(text):
    """return the number of letters in the text"""
    # Using regex, find the number of letters in the text.
    letters = 0
    for i in range(len(text)):
        if re.match("[a-zA-Z]", text[i]):
            letters += 1

    return letters


# return the number of words in the text
def count_words(text):
    """return the number of words in the text"""
    return len(text.split())


# return the number of sentences in the text
def count_sentences(text):
    """return the number of sentences in the text"""
    return text.count(".") + text.count("!") + text.count("?")


# compute the colman-liau index
def coleman_liau_index(text):
    """compute the colman-liau index"""
    L = count_letters(text) / count_words(text) * 100
    S = count_sentences(text) / count_words(text) * 100
    index = 0.0588 * L - 0.296 * S - 15.8
    # Round index to the nearest whole number
    index = round(index)
    return index

# if the index is less than 1, print "Before Grade 1"
# if the index is between 1 and 16, print "Grade {0}"
# if the index is greater than 16, print "Grade 16+"
def print_grade(index):
    """print the grade of the text"""
    if index < 1:
        print("Before Grade 1")
    elif index >= 1 and index <= 16:
        print("Grade {0}".format(int(index)))
    else:
        print("Grade 16+")


# print the index
print_grade(coleman_liau_index(TEXT))
